package/MDAnalysis/analysis/lineardensity.py

```diff
@@ -199,10 +210,24 @@ def _conclude(self):
             for key in ['pos', 'pos_std', 'char', 'char_std']:
                 self.results[dim][key] /= self.n_frames
             # Compute standard deviation for the error
-            self.results[dim]['pos_std'] = np.sqrt(self.results[dim][
-                'pos_std'] - np.square(self.results[dim]['pos']))
-            self.results[dim]['char_std'] = np.sqrt(self.results[dim][
-                'char_std'] - np.square(self.results[dim]['char']))
+            # For certain tests in testsuite, floating point imprecision
+            # can lead to negative radicands of tiny magnitude (yielding nan),
+            # rather than actual value of zero. radicand_pos and radicand_char
+            # are therefore tested for closeness to 0 first and changed where
+            # appropriate before the square root is calculated.
+            radicand_pos = self.results[dim][
+                'pos_std'] - np.square(self.results[dim]['pos'])
+            for i in range(len(radicand_pos)):
+                if np.isclose(radicand_pos, np.zeros(self.nbins))[i]:
+                    radicand_pos[i] = 0
```

Comment:
```suggestion
            radicand_pos = np.where(np.isclose(radicand_pos, 0), 0, radicand_pos)
```
or something on those lines.

Reply:
Or even
```suggestion
            radicand_pos[radicand_pos < 0] = 0
```

Reply:
Thanks @RMeli! I was sure there must be a better solution for this and your suggestion looks much cleaner. 
I think using your first suggestion with `np.where` has two advantages: 

- If an actual negative number that's not close to zero shows up at this point, that's an error that should be flagged up/ throw a warning
- positive numbers on the order of e-14 are not meaningful results and should be 0 as well

But your second suggestion looks like it would give better performance. I'm not sure how much of a difference this would make, could people weigh in here? If the first suggestion is not much slower than the second, than that's the one I would go for.

Reply:
Yes, precisely. The first option is safer, the second one is faster. 
```
In [1]: import numpy as np

In [2]: x = np.array([1.0, 1e-10, -1e-10, 0.1])

In [3]: %timeit y = np.where(np.isclose(x, 0), 0, x)
40.4 µs ± 325 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: z = x.copy()

In [5]: %timeit z[z < 0] = 0
2.1 µs ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

```
In [6]: x = np.random.normal(size=10000)

In [7]: %timeit y = np.where(np.isclose(x, 0), 0, x)
90.5 µs ± 1.23 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: z = x.copy()

In [9]: %timeit z[z < 0] = 0
6.72 µs ± 65.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

In scientific software, IMHO, correctness is far more important than speed. However, since this is the `_conclude` function, if the code executed before performs enough checks it might be safe to use the second option here (i.e. the only negative values only come from numerical errors).

Reply:
Because we only have to do this once, I might vote for the safer option here, unless we can guarantee that negative numbers are always floating point issues (I've not really delved into what the cause is here, maybe that is the case).

Reply:
Thanks for this clear breakdown, RMeli! 

IAlibay, I have looked into the code again and I can't see a way for actual negative numbers to appear here. At least not unless things really break, which should be caught by the tests. I would suggest the following, then:

`radicand_pos[radicand_pos < 1e-9] = 0`

This is fast and also takes care of floating point errors that give a tiny positive number. 

`x = np.array([1.0, 1e-10, -1e-10, 0.1])`
`%timeit x[x < 1e-9] = 0`
`1.07 µs ± 8.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)`

Reply:
Yeah sure I can live with that

Reply:
Changed now.

Reply:
Sorry for the drive-by comment: If you're only concerned with negative numbers then I'd keep @RMeli 's suggestion with `x[x<0]` instead of introducing an arbitrary cutoff (and 1e-9 is not necessarily small). Numerical roundoff is normal and if anything I'd want to take care of it at the very end of a calculation and not infer in the middle if something ought to be zero or not. The objective of "no negative numbers" is clear. The objective of "should be zero" is not. Thus I only would implement the first one.

Reply:
I am confused by his operation. Do have an example where this happens? I never had an issue with `nan` in calculating density profiles.

Reply:
That's a good point, @orbeckst, thanks! I have changed this to `x[x<0]` now. 

@PicoCentauri, Ah, I did not know about `np.nan_to_num`, thank you! For now, I stuck with the other suggestion because it is slightly faster, but is there another advantage to nan_to_num?

Reply:
If it is faster it is always better. I am not aware of any other advantages compared to your solution.